### ISSUE TITLE
ReformerEncDec generate slow function 

### ISSUE THREAD

**Issue by allhelllooz** [lucidrains/reformer-pytorch#114]
I am trying to run following code ... similar to the code given in ReadMe file.
You can see that generating 1 sample took >30 sec.
And training just 1 encoder-decoder sample took ~2 sec.
It runs on gpu .. but feels super slow.
Is there a problem with code or example given ??
Checked gpu .. it is getting used ~3.5gb when I run the code.

System - Windows 10, Anaconda python 3.6, torch 1.4

Code - 
```
import torch
import pycuda.driver as cuda
cuda.init()
torch.cuda.current_device()
cuda.Device(0).name()

from reformer_pytorch import ReformerEncDec
from datetime import datetime


t1 = datetime.now()
print(t1)
DE_SEQ_LEN = 1024
EN_SEQ_LEN = 256

enc_dec = ReformerEncDec(
    dim = 512,
    enc_num_tokens = 64000,
    enc_depth = 4,
    enc_max_seq_len = DE_SEQ_LEN,
    dec_num_tokens = 64000,
    dec_depth = 4,
    dec_max_seq_len = EN_SEQ_LEN
).cuda()

t2 = datetime.now()
print("Stage1 time : ",(t2-t1).total_seconds())
train_seq_in = torch.randint(0, 64000, (1, DE_SEQ_LEN)).long().cuda()
train_seq_out = torch.randint(0, 64000, (1, EN_SEQ_LEN)).long().cuda()
input_mask = torch.ones(1, DE_SEQ_LEN).bool().cuda()

loss = enc_dec(train_seq_in, train_seq_out, return_loss = True, enc_input_mask = input_mask)
loss.backward()
# learn

t3 = datetime.now()
print("Stage2 time : ",(t3-t2).total_seconds())
# evaluate with the following
eval_seq_in = torch.randint(0, 64000, (1, DE_SEQ_LEN)).long().cuda()
eval_seq_out_start = torch.tensor([[0.]]).long().cuda() # assume 0 is id of start token
samples = enc_dec.generate(eval_seq_in, eval_seq_out_start, seq_len = EN_SEQ_LEN, eos_token = 1) # assume 1 is id of stop token
print(samples.shape) # (1, <= 1024) decode the tokens

t4 = datetime.now()
print("Stage3 time : ",(t4-t3).total_seconds())

```

Output
```
'GeForce GTX 1070 with Max-Q Design'
2020-09-22 15:08:09.806857
Stage1 time :  1.995265
Stage2 time :  2.430372
torch.Size([1, 256])
Stage3 time :  33.585908

```

**Comment by lucidrains** [comment]
@allhelllooz that's unfortunately the cost that comes with the Reformer. the reversibility, LSH, and hashing rounds makes this technique slower than others. you are always free to try out some of the alternative sparse attention architectures, which I have many! (sinkhorn, linear, routing)

**Comment by allhelllooz** [comment]
Sure. Thanks for the reply. Do you have any kind of benchmark done for speed vs accuracy for these transformers ? @lucidrains 
Less memory is an advantage but more time in inference is not.

**Comment by lucidrains** [comment]
@allhelllooz each of the repositories come with a working example for enwik8, so you could explore the hyperparameter space and get a sense of the tradeoffs. I think, for a full encoder / decoder, I would go with either Routing Transformer https://github.com/lucidrains/routing-transformer or Linear Attention Transformer https://github.com/lucidrains/linear-attention-transformer. Sinkhorn is also worth trying, but comes with generalization problems for the decoder that can only be overcome with special training https://github.com/lucidrains/sinkhorn-transformer#decoding-and-sequence-lengths If you desire only auto-regressive (GPT-like), then you can also try out Compressive Transformers https://github.com/lucidrains/compressive-transformer-pytorch

**Comment by allhelllooz** [comment]
Cool. Thanks for the reply.
My major focus is on encode-decoder ... seq2seq & summarization. Was working with ProphetNet & pegasus .. and found both of them very memory intensive. They probably need ~24gb to even run with decent batch size. 
Will try out what you suggested. Closing the issue. Thanks.
